File: Sources/Infrastructure/Statistics/SQLiteTrafficLedger.swift
import Foundation

actor SQLiteTrafficLedger: TrafficLedgerStoring {
  private let databaseURL: URL
  private var persistence: TrafficLedgerPersistence?
  private var runtimeState = TrafficLedgerRuntimeState()
  private var isPrepared = false
  private var lastPrunedLocalDay: String?

  init(databaseURL: URL) {
    self.databaseURL = databaseURL
  }

  func prepare(calendar: Calendar, now: Date) async throws -> TrafficStatisticsSnapshot {
    let persistence = try requirePersistence()
    try prepareIfNeeded(calendar: calendar, now: now, persistence: persistence)
    return try snapshot(calendar: calendar, now: now)
  }

  func beginMonitoring(
    version: String,
    at date: Date,
    calendar: Calendar
  ) async throws -> TrafficStatisticsSnapshot {
    let persistence = try preparedPersistence(calendar: calendar, now: date)
    try persistence.transaction {
      if runtimeState.currentSessionID == nil {
        try openCoreSession(version: version, at: date, persistence: persistence)
      } else if runtimeState.currentMihomoVersion != version {
        try replaceCoreSession(
          version: version,
          at: date,
          reason: "version_change",
          persistence: persistence
        )
      }
    }
    return try snapshot(calendar: calendar, now: date)
  }

  func record(
    _ observation: TrafficLedgerObservation,
    calendar: Calendar
  ) async throws -> TrafficStatisticsSnapshot {
    let persistence = try preparedPersistence(calendar: calendar, now: observation.observedAt)
    try persistence.transaction {
      let opensFreshSession = runtimeState.currentSessionID == nil
      if runtimeState.currentSessionID == nil {
        try openCoreSession(
          version: runtimeState.currentMihomoVersion ?? "unknown",
          at: observation.observedAt,
          persistence: persistence
        )
      }

      if !opensFreshSession {
        switch observation.transition {
        case .baselineEstablished:
          try recordReconnectGapIfNeeded(
            currentKernelTotal: observation.kernelTotal,
            observedAt: observation.observedAt,
            calendar: calendar,
            persistence: persistence
          )
        case .delta(let report):
          try add(
            report.categories,
            observedAt: observation.observedAt,
            calendar: calendar,
            persistence: persistence
          )
        case .countersReset:
          try replaceCoreSession(
            version: runtimeState.currentMihomoVersion ?? "unknown",
            at: observation.observedAt,
            reason: "counter_reset",
            persistence: persistence
          )
        }
      }

      runtimeState.lastObservedAt = observation.observedAt
      runtimeState.lastKernelTotal = observation.kernelTotal
      try persistence.saveRuntimeState(runtimeState)
      try pruneIfNeeded(
        calendar: calendar,
        now: observation.observedAt,
        persistence: persistence
      )
    }
    return try snapshot(calendar: calendar, now: observation.observedAt)
  }

  func startInterval(
    name: String,
    note: String?,
    at date: Date,
    calendar: Calendar
  ) async throws -> TrafficStatisticsSnapshot {
    let persistence = try preparedPersistence(calendar: calendar, now: date)
    let normalizedName = try TrafficIntervalInput.normalizedName(name)
    try persistence.transaction {
      try persistence.intervals.insert(
        id: UUID(),
        name: normalizedName,
        note: TrafficIntervalInput.normalizedNote(note),
        startedAt: date,
        baseline: try persistence.totals().proxy
      )
    }
    return try snapshot(calendar: calendar, now: date)
  }

  func stopInterval(
    id: UUID,
    at date: Date,
    calendar: Calendar
  ) async throws -> TrafficStatisticsSnapshot {
    let persistence = try preparedPersistence(calendar: calendar, now: date)
    try persistence.transaction {
      try persistence.intervals.complete(
        id: id,
        endedAt: date,
        baseline: try persistence.totals().proxy
      )
    }
    return try snapshot(calendar: calendar, now: date)
  }

  func renameInterval(
    id: UUID,
    name: String,
    calendar: Calendar,
    now: Date
  ) async throws -> TrafficStatisticsSnapshot {
    let persistence = try preparedPersistence(calendar: calendar, now: now)
    try persistence.transaction {
      try persistence.intervals.rename(
        id: id,
        name: try TrafficIntervalInput.normalizedName(name)
      )
    }
    return try snapshot(calendar: calendar, now: now)
  }

  func deleteInterval(
    id: UUID,
    calendar: Calendar,
    now: Date
  ) async throws -> TrafficStatisticsSnapshot {
    let persistence = try preparedPersistence(calendar: calendar, now: now)
    try persistence.transaction {
      try persistence.intervals.delete(id: id)
    }
    return try snapshot(calendar: calendar, now: now)
  }

  func interruptActiveIntervals(
    reason: TrafficIntervalEndReason,
    calendar: Calendar,
    now: Date
  ) async throws -> TrafficStatisticsSnapshot {
    let persistence = try preparedPersistence(calendar: calendar, now: now)
    try persistence.transaction {
      let interruptedAt = runtimeState.lastObservedAt ?? now
      try persistence.intervals.interruptActive(
        endedAt: interruptedAt,
        baseline: try persistence.totals().proxy,
        reason: reason
      )
      if let sessionID = runtimeState.currentSessionID {
        try persistence.closeCoreSession(
          id: sessionID,
          endedAt: interruptedAt,
          reason: reason.rawValue
        )
      }
      runtimeState.currentSessionID = nil
      runtimeState.currentMihomoVersion = nil
      runtimeState.lastKernelTotal = nil
      try persistence.saveRuntimeState(runtimeState)
    }
    return try snapshot(calendar: calendar, now: now)
  }

  func clear(calendar: Calendar, now: Date) async throws -> TrafficStatisticsSnapshot {
    let persistence = try requirePersistence()
    try persistence.reset()
    runtimeState = TrafficLedgerRuntimeState()
    isPrepared = true
    lastPrunedLocalDay = nil
    return try snapshot(calendar: calendar, now: now)
  }

  private func requirePersistence() throws -> TrafficLedgerPersistence {
    if let persistence {
      return persistence
    }
    let persistence = try TrafficLedgerPersistence(databaseURL: databaseURL)
    self.persistence = persistence
    return persistence
  }

  private func preparedPersistence(
    calendar: Calendar,
    now: Date
  ) throws -> TrafficLedgerPersistence {
    let persistence = try requirePersistence()
    try prepareIfNeeded(calendar: calendar, now: now, persistence: persistence)
    return persistence
  }

  private func prepareIfNeeded(
    calendar: Calendar,
    now: Date,
    persistence: TrafficLedgerPersistence
  ) throws {
    guard !isPrepared else {
      return
    }
    runtimeState = try persistence.loadRuntimeState()
    try persistence.transaction {
      let lifetime = try persistence.totals()
      let interruptedAt = runtimeState.lastObservedAt ?? now
      try persistence.intervals.interruptActive(
        endedAt: interruptedAt,
        baseline: lifetime.proxy,
        reason: .recovery
      )
      if let sessionID = runtimeState.currentSessionID {
        try persistence.closeCoreSession(
          id: sessionID,
          endedAt: interruptedAt,
          reason: TrafficIntervalEndReason.recovery.rawValue
        )
      }
      runtimeState.currentSessionID = nil
      runtimeState.currentMihomoVersion = nil
      runtimeState.lastKernelTotal = nil
      try persistence.saveRuntimeState(runtimeState)
      try pruneIfNeeded(calendar: calendar, now: now, persistence: persistence)
    }
    isPrepared = true
  }

  private func openCoreSession(
    version: String,
    at date: Date,
    persistence: TrafficLedgerPersistence
  ) throws {
    let sessionID = UUID()
    try persistence.createCoreSession(id: sessionID, version: version, startedAt: date)
    runtimeState.currentSessionID = sessionID
    runtimeState.currentMihomoVersion = version
    runtimeState.lastKernelTotal = nil
    try persistence.saveRuntimeState(runtimeState)
  }

  private func replaceCoreSession(
    version: String,
    at date: Date,
    reason: String,
    persistence: TrafficLedgerPersistence
  ) throws {
    if let sessionID = runtimeState.currentSessionID {
      try persistence.closeCoreSession(id: sessionID, endedAt: date, reason: reason)
    }
    try openCoreSession(version: version, at: date, persistence: persistence)
  }

  private func recordReconnectGapIfNeeded(
    currentKernelTotal: TrafficBytes,
    observedAt: Date,
    calendar: Calendar,
    persistence: TrafficLedgerPersistence
  ) throws {
    guard let previousKernelTotal = runtimeState.lastKernelTotal else {
      return
    }
    guard
      let gap = TrafficBytes.nonnegativeDelta(
        current: currentKernelTotal,
        previous: previousKernelTotal
      )
    else {
      try replaceCoreSession(
        version: runtimeState.currentMihomoVersion ?? "unknown",
        at: observedAt,
        reason: "counter_reset",
        persistence: persistence
      )
      return
    }
    try add(
      CategorizedTrafficBytes.zero.adding(gap, to: .unknown),
      observedAt: observedAt,
      calendar: calendar,
      persistence: persistence
    )
  }

  private func add(
    _ categories: CategorizedTrafficBytes,
    observedAt: Date,
    calendar: Calendar,
    persistence: TrafficLedgerPersistence
  ) throws {
    guard let sessionID = runtimeState.currentSessionID else {
      throw TrafficStatisticsError.database("缺少内核统计会话")
    }
    try persistence.add(
      categories,
      observedAt: observedAt,
      calendar: calendar,
      coreSessionID: sessionID
    )
  }

  private func pruneIfNeeded(
    calendar: Calendar,
    now: Date,
    persistence: TrafficLedgerPersistence
  ) throws {
    let localDay = TrafficLedgerPersistence.localDay(for: now, calendar: calendar)
    guard localDay != lastPrunedLocalDay else {
      return
    }
    guard let cutoff = calendar.date(byAdding: .day, value: -365, to: now) else {
      throw TrafficStatisticsError.database("无法计算分钟桶保留时间")
    }
    try persistence.pruneBuckets(before: cutoff)
    lastPrunedLocalDay = localDay
  }

  private func snapshot(calendar: Calendar, now: Date) throws -> TrafficStatisticsSnapshot {
    let persistence = try requirePersistence()
    let lifetime = try persistence.totals()
    let localDay = TrafficLedgerPersistence.localDay(for: now, calendar: calendar)
    let recentLocalDays = try recentLocalDays(calendar: calendar, now: now)
    let storedDays = try persistence.dailyTotals(
      category: .proxy,
      since: recentLocalDays.first ?? localDay
    )
    let storedByDay = Dictionary(uniqueKeysWithValues: storedDays.map { ($0.localDay, $0) })
    return TrafficStatisticsSnapshot(
      today: try persistence.totals(localDay: localDay),
      lifetime: lifetime,
      intervals: try persistence.intervals.load(currentProxyTotal: lifetime.proxy),
      recentProxyDays: recentLocalDays.map {
        storedByDay[$0] ?? TrafficDailyTotal(localDay: $0, bytes: .zero)
      },
      lastObservedAt: runtimeState.lastObservedAt
    )
  }

  private func recentLocalDays(calendar: Calendar, now: Date) throws -> [String] {
    try (0..<30).reversed().map { offset in
      guard let date = calendar.date(byAdding: .day, value: -offset, to: now) else {
        throw TrafficStatisticsError.database("无法计算最近每日统计日期")
      }
      return TrafficLedgerPersistence.localDay(for: date, calendar: calendar)
    }
  }
}

File: platform/windows/MihomoMeter.Windows.Core/Infrastructure/Statistics/SQLiteTrafficLedger.cs
using Microsoft.Data.Sqlite;
using MihomoMeter.Windows.Core.Application;
using MihomoMeter.Windows.Core.Domain;

namespace MihomoMeter.Windows.Core.Infrastructure.Statistics;

public sealed class SQLiteTrafficLedger : ITrafficLedger
{
    private readonly string _databasePath;
    private readonly SemaphoreSlim _gate = new(1, 1);
    private TrafficLedgerPersistence? _persistence;
    private TrafficLedgerRuntimeState _runtimeState = new();
    private bool _isPrepared;
    private string? _lastPrunedLocalDay;

    public SQLiteTrafficLedger(string databasePath)
    {
        _databasePath = databasePath;
    }

    public Task<TrafficStatisticsSnapshot> PrepareAsync(
        TimeZoneInfo timeZone,
        DateTimeOffset now,
        CancellationToken cancellationToken)
    {
        return ExecuteAsync(() =>
        {
            var persistence = RequirePersistence();
            PrepareIfNeeded(timeZone, now, persistence);
            return Snapshot(timeZone, now, persistence);
        }, cancellationToken);
    }

    public Task<TrafficStatisticsSnapshot> BeginMonitoringAsync(
        string version,
        TimeZoneInfo timeZone,
        DateTimeOffset now,
        CancellationToken cancellationToken)
    {
        return ExecuteAsync(() =>
        {
            var persistence = PreparedPersistence(timeZone, now);
            persistence.Transaction(transaction =>
            {
                if (_runtimeState.CurrentSessionId is null)
                {
                    OpenCoreSession(version, now, persistence, transaction);
                }
                else if (!string.Equals(
                    _runtimeState.CurrentMihomoVersion,
                    version,
                    StringComparison.Ordinal))
                {
                    ReplaceCoreSession(
                        version,
                        now,
                        "version_change",
                        persistence,
                        transaction);
                }
            });
            return Snapshot(timeZone, now, persistence);
        }, cancellationToken);
    }

    public Task<TrafficStatisticsSnapshot> RecordAsync(
        TrafficLedgerObservation observation,
        TimeZoneInfo timeZone,
        CancellationToken cancellationToken)
    {
        return ExecuteAsync(() =>
        {
            var persistence = PreparedPersistence(timeZone, observation.ObservedAt);
            persistence.Transaction(transaction =>
            {
                var opensFreshSession = _runtimeState.CurrentSessionId is null;
                if (opensFreshSession)
                {
                    OpenCoreSession(
                        _runtimeState.CurrentMihomoVersion ?? "unknown",
                        observation.ObservedAt,
                        persistence,
                        transaction);
                }

                if (!opensFreshSession)
                {
                    switch (observation.Transition)
                    {
                        case TrafficLedgerBaselineEstablished:
                            RecordReconnectGapIfNeeded(
                                observation.KernelTotal,
                                observation.ObservedAt,
                                timeZone,
                                persistence,
                                transaction);
                            break;
                        case TrafficLedgerDelta delta:
                            Add(
                                delta.Report.Categories,
                                observation.ObservedAt,
                                timeZone,
                                persistence,
                                transaction);
                            break;
                        case TrafficLedgerCountersReset:
                            ReplaceCoreSession(
                                _runtimeState.CurrentMihomoVersion ?? "unknown",
                                observation.ObservedAt,
                                "counter_reset",
                                persistence,
                                transaction);
                            break;
                        default:
                            throw new TrafficStatisticsException("本地统计观测状态无效。");
                    }
                }

                _runtimeState = _runtimeState with
                {
                    LastObservedAt = observation.ObservedAt,
                    LastKernelTotal = observation.KernelTotal,
                };
                persistence.SaveRuntimeState(_runtimeState, transaction);
                PruneIfNeeded(timeZone, observation.ObservedAt, persistence, transaction);
            });
            return Snapshot(timeZone, observation.ObservedAt, persistence);
        }, cancellationToken);
    }

    public Task<TrafficStatisticsSnapshot> InterruptMonitoringAsync(
        TrafficSessionEndReason reason,
        TimeZoneInfo timeZone,
        DateTimeOffset now,
        CancellationToken cancellationToken)
    {
        return ExecuteAsync(() =>
        {
            var persistence = PreparedPersistence(timeZone, now);
            persistence.Transaction(transaction =>
            {
                var endedAt = _runtimeState.LastObservedAt ?? now;
                if (_runtimeState.CurrentSessionId is Guid sessionId)
                {
                    persistence.CloseCoreSession(
                        sessionId,
                        endedAt,
                        EndReasonName(reason),
                        transaction);
                }

                _runtimeState = _runtimeState with
                {
                    CurrentSessionId = null,
                    CurrentMihomoVersion = null,
                    LastKernelTotal = null,
                };
                persistence.SaveRuntimeState(_runtimeState, transaction);
            });
            return Snapshot(timeZone, now, persistence);
        }, cancellationToken);
    }

    public async ValueTask DisposeAsync()
    {
        await _gate.WaitAsync().ConfigureAwait(false);
        try
        {
            _persistence?.Dispose();
            _persistence = null;
        }
        finally
        {
            _gate.Release();
            _gate.Dispose();
        }
    }

    private async Task<TrafficStatisticsSnapshot> ExecuteAsync(
        Func<TrafficStatisticsSnapshot> operation,
        CancellationToken cancellationToken)
    {
        await _gate.WaitAsync(cancellationToken).ConfigureAwait(false);
        try
        {
            cancellationToken.ThrowIfCancellationRequested();
            return operation();
        }
        catch (OperationCanceledException)
        {
            throw;
        }
        catch (TrafficStatisticsException)
        {
            throw;
        }
        catch (Exception exception) when (
            exception is SqliteException
                or IOException
                or UnauthorizedAccessException
                or OverflowException
                or FormatException)
        {
            throw new TrafficStatisticsException("本地统计数据库暂不可用。", exception);
        }
        finally
        {
            _gate.Release();
        }
    }

    private TrafficLedgerPersistence RequirePersistence()
    {
        if (_persistence is not null)
        {
            return _persistence;
        }

        _persistence = new TrafficLedgerPersistence(_databasePath);
        return _persistence;
    }

    private TrafficLedgerPersistence PreparedPersistence(
        TimeZoneInfo timeZone,
        DateTimeOffset now)
    {
        var persistence = RequirePersistence();
        PrepareIfNeeded(timeZone, now, persistence);
        return persistence;
    }

    private void PrepareIfNeeded(
        TimeZoneInfo timeZone,
        DateTimeOffset now,
        TrafficLedgerPersistence persistence)
    {
        if (_isPrepared)
        {
            return;
        }

        _runtimeState = persistence.LoadRuntimeState();
        persistence.Transaction(transaction =>
        {
            var interruptedAt = _runtimeState.LastObservedAt ?? now;
            if (_runtimeState.CurrentSessionId is Guid sessionId)
            {
                persistence.CloseCoreSession(
                    sessionId,
                    interruptedAt,
                    EndReasonName(TrafficSessionEndReason.Recovery),
                    transaction);
            }

            _runtimeState = _runtimeState with
            {
                CurrentSessionId = null,
                CurrentMihomoVersion = null,
                LastKernelTotal = null,
            };
            persistence.SaveRuntimeState(_runtimeState, transaction);
            PruneIfNeeded(timeZone, now, persistence, transaction);
        });
        _isPrepared = true;
    }

    private void OpenCoreSession(
        string version,
        DateTimeOffset date,
        TrafficLedgerPersistence persistence,
        SqliteTransaction transaction)
    {
        var sessionId = Guid.NewGuid();
        persistence.CreateCoreSession(sessionId, version, date, transaction);
        _runtimeState = _runtimeState with
        {
            CurrentSessionId = sessionId,
            CurrentMihomoVersion = version,
            LastKernelTotal = null,
        };
        persistence.SaveRuntimeState(_runtimeState, transaction);
    }

    private void ReplaceCoreSession(
        string version,
        DateTimeOffset date,
        string reason,
        TrafficLedgerPersistence persistence,
        SqliteTransaction transaction)
    {
        if (_runtimeState.CurrentSessionId is Guid sessionId)
        {
            persistence.CloseCoreSession(sessionId, date, reason, transaction);
        }

        OpenCoreSession(version, date, persistence, transaction);
    }

    private void RecordReconnectGapIfNeeded(
        TrafficBytes currentKernelTotal,
        DateTimeOffset observedAt,
        TimeZoneInfo timeZone,
        TrafficLedgerPersistence persistence,
        SqliteTransaction transaction)
    {
        if (_runtimeState.LastKernelTotal is not TrafficBytes previousKernelTotal)
        {
            return;
        }

        var gap = TrafficBytes.NonnegativeDelta(currentKernelTotal, previousKernelTotal);
        if (gap is null)
        {
            ReplaceCoreSession(
                _runtimeState.CurrentMihomoVersion ?? "unknown",
                observedAt,
                "counter_reset",
                persistence,
                transaction);
            return;
        }

        Add(
            CategorizedTrafficBytes.Zero.Adding(gap.Value, TrafficCategory.Unknown),
            observedAt,
            timeZone,
            persistence,
            transaction);
    }

    private void Add(
        CategorizedTrafficBytes categories,
        DateTimeOffset observedAt,
        TimeZoneInfo timeZone,
        TrafficLedgerPersistence persistence,
        SqliteTransaction transaction)
    {
        if (_runtimeState.CurrentSessionId is not Guid sessionId)
        {
            throw new TrafficStatisticsException("缺少本地统计内核会话。");
        }

        persistence.Add(
            categories,
            observedAt,
            timeZone,
            sessionId,
            transaction);
    }

    private void PruneIfNeeded(
        TimeZoneInfo timeZone,
        DateTimeOffset now,
        TrafficLedgerPersistence persistence,
        SqliteTransaction transaction)
    {
        var localDay = TrafficLedgerPersistence.LocalDay(now, timeZone);
        if (string.Equals(localDay, _lastPrunedLocalDay, StringComparison.Ordinal))
        {
            return;
        }

        persistence.PruneBuckets(now.AddDays(-365), transaction);
        _lastPrunedLocalDay = localDay;
    }

    private TrafficStatisticsSnapshot Snapshot(
        TimeZoneInfo timeZone,
        DateTimeOffset now,
        TrafficLedgerPersistence persistence)
    {
        return new TrafficStatisticsSnapshot(
            persistence.Totals(TrafficLedgerPersistence.LocalDay(now, timeZone)),
            persistence.Totals(),
            _runtimeState.LastObservedAt);
    }

    private static string EndReasonName(TrafficSessionEndReason reason)
    {
        return reason switch
        {
            TrafficSessionEndReason.MonitoringStopped => "monitoring_stopped",
            TrafficSessionEndReason.ApplicationExit => "application_exit",
            TrafficSessionEndReason.TerminalFailure => "terminal_failure",
            TrafficSessionEndReason.Recovery => "recovery",
            _ => throw new TrafficStatisticsException("本地统计结束原因无效。"),
        };
    }
}
